src/cpp/session/modules/SessionConnectionsInstaller.R

```diff
@@ -0,0 +1,476 @@
+})
+
+.rs.addFunction("odbcBundleCheckPrereqsBrew", function() {
+   identical(system2("brew", stdout = FALSE), 1L)
+})
+
+.rs.addFunction("odbcBundleCheckPrereqsOsx", function() {
+   if (!.rs.odbcBundleCheckPrereqsUnixodbc()) {
+      if (!.rs.odbcBundleCheckPrereqsBrew()) {
+         message("Installing Brew...")
+         system2(
+            "/usr/bin/ruby",
+            args = list(
+               "-e",
+               "\"$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)\""
+            )
+         )
+      }
+      
+      message("Installing unixODBC...")
+      system2(
+         "brew",
+         args = list("install", "unixodbc")
+      )
+   }
+   
+   if (!.rs.odbcBundleCheckPrereqsUnixodbc())
+      stop("Failed to install unixODBC, please install from www.unixodbc.org")
+})
+
+.rs.addFunction("odbcBundleCheckPrereqsLinux", function() {
+   if (!.rs.odbcBundleCheckPrereqsUnixodbc())
+      stop("unixODBC is not installed, please install from www.unixodbc.org")
+})
+
+.rs.addFunction("odbcBundleRegistryAdd", function(entries) {
+   validateEntry <- function(entry) {
+      tryCatch({
+         verify <- readRegistry(entry$path, "HLM")
+         identical(verify[[entry$key]], entry$value)
+      }, error = function(e) {
+         FALSE
+      })
+   }
+
+   odbcFileEscape <- function(value) {
+    gsub("\\\\", "\\\\\\\\", value)
+   }
+
+   if (all(sapply(entries, function(e) validateEntry(e))))
+      return()
+
+   allAdded <- TRUE
+   for (entry in entries) {
+     fullPath <- file.path("HKEY_LOCAL_MACHINE", entry$path, fsep = "\\")
+     system2(
+        "REG",
+        args = list(
+           "ADD",
+           shQuote(fullPath),
+           "/v",
+           shQuote(entry$key),
+           "/t",
+           "REG_SZ",
+           "/d",
+           shQuote(entry$value),
+           "/f"
+        )
+     )
+
+     if (!validateEntry(entry)) {
+      allAdded <- FALSE
+      break
+     }
+   }
+
+   if (!allAdded) {
+      message("Could not add registry keys from R, retrying using registry prompt.")
+      addReg <- tempfile(fileext = ".reg")
+
+      lineEntries <- sapply(entries, function(entry) {
+         fullPath <- file.path("HKEY_LOCAL_MACHINE", entry$path, fsep = "\\")
+         c(
+            paste("[", fullPath, "]", sep = ""),
+            paste(
+               "\"",
+               entry$key,
+               "\"=\"",
+               odbcFileEscape(entry$value),
+               "\"",
+               sep = ""
+            ),
+            ""
+         )
+      })
+
+      lines <- c(
+         "REGEDIT4",
+         "",
+         lineEntries
+      )
+
+      writeLines(lines, addReg)
+
+      message("Waiting for ", addReg, " to be registered.")
+      system2(
+         "explorer",
+         addReg
+      )
+
+      allEntriesValid <- function() {
+        all(sapply(entries, function(e) validateEntry(e)))
+      }
+
+      registryStart <- Sys.time()
+      registryWait <- 300
```

Comment:
I think we should give up sooner here (after 60 seconds or so unless you think it'll take longer), and print some minimal progress once a second while waiting, even if it's just periods. Users are probably going to kill this if it goes radio silent for more than 30 seconds without printing anything.